authentication/certificate-types-descriptions.adoc

```diff
@@ -0,0 +1,47 @@
+[id="ocp-certificates"]
+= Certificate types and descriptions
+include::modules/common-attributes.adoc[]
+:context: ocp-certificates
+
+toc::[]
+
+== Certificate validation
+
+{product-title} monitors certificates for proper validity, for the cluster
+certificates it issues and manages. The {product-title} alerting framework has
+rules to help identify when a certificate issue is about to occur. These rules
+consist of the following checks:
+
+* API server client certificate expiration is less than five minutes.
+
+include::modules/user-provided-certificates-for-api-server.adoc[leveloffset=+1]
+include::modules/proxy-certificates.adoc[leveloffset=+1]
+include::modules/ingress-certificates.adoc[leveloffset=+1]
+include::modules/user-provided-certificates-for-default-ingress.adoc[leveloffset=+1]
+include::modules/service-ca-certificates.adoc[leveloffset=+1]
+include::modules/node-certificates.adoc[leveloffset=+1]
+include::modules/bootstrap-certificates.adoc[leveloffset=+1]
+include::modules/etcd-certificates.adoc[leveloffset=+1]
+
+== Monitoring and cluster logging Operator component certificates
+
+Monitoring components secure their traffic with service CA certificates. These
+certificates are valid for 2 years and are replaced automatically on rotation of
+the service CA, which is every 13 months.
+
+[discrete]
+== Management
+
+These certificates are managed by the system and not the user.
```

Comment:
@s-urbaniak can we say here, that if it lives in `openshift-monitoring` it is system managed and rotated automatically?

@jcantrill same question but for logging namespaces.

Reply:
sounds like @jcantrill confirmed this in our call today for the logging namespace

Reply:
@pweil hrm, somehow github i missed this github notification :-/ yes, internal traffic service CA certificates are rotated by the system.

Reply:
No worries, I'm the same way with GH alerts and BZ alerts.  Too many to NOT miss at least one.  Thanks for confirming.

@ahardin-rh looks like we can add that update for the `openshift-monitoring` namespace.  Thanks!

Reply:
Reconcile() calls CreateOrUpdateLogStore().
https://github.com/openshift/cluster-logging-operator/blob/01bf653aa2290b0230d164daa255d0c58c5f1a92/pkg/k8shandler/reconciler.go#L38

CreateOrUpdateLogStore() runs cert_generation.sh.
https://github.com/openshift/cluster-logging-operator/blob/57b091eb6ef0d1f2e263d5aa70e7d3e1b4666390/pkg/k8shandler/certificates.go#L127

cert_generation.sh checks the expire dates by using "openssl x509 -checkend 0" command, and regenerate certificates if they have already expired.
https://github.com/openshift/cluster-logging-operator/blob/ec11d6a5fe5f3c01c2793fc19793172bf51ad15b/scripts/cert_generation.sh

I confirmed certificates are regenerated as @jcantrill said.

However, I have the following concerns.
- Pods will need to be recreated in order to reload the regenerated certificates. So, who will recreate them?
- cert_generation.sh regenerates certificates just after they expired. On the other hand, other certificates(e.g. service-ca certificate) are regenarated before a few month ago from the expiration date. cert_generation.sh also should regenerate before a few month ago.

Reply:
> Pods will need to be recreated in order to reload the regenerated certificates. So, who will recreate them?

For openshift-monitoring, It looks like the same problem will occur.
https://github.com/openshift/cluster-monitoring-operator/blob/0a458269aecb799b162aa9dc8fc61ccc0fa2be1b/pkg/manifests/tls.go#L68

I would appreciate if you can help to clear my concerns.

Reply:
> Pods will need to be recreated in order to reload the regenerated certificates. So, who will recreate them?

For openshift-monitoring, It looks like the same problem will occur.
https://github.com/openshift/cluster-monitoring-operator/blob/0a458269aecb799b162aa9dc8fc61ccc0fa2be1b/pkg/manifests/tls.go#L68

I would appreciate if you can help to clear my concerns.

Reply:
> Pods will need to be recreated in order to reload the regenerated certificates. So, who will recreate them?

For cluster-logging, it regenerates certificates by using cert_generation.sh.
So I tested the regeneration by modifying cert_generation.sh.

As a result:
- ElasticSearch pod is recreated automatically after certificates are generated
- Kibana's certificates are reloaded without recreating pod
- Fluentd's certificates aren't reloaded. It's a problem
- Become unable to access kibana route since the route has old certificate. It's also the problem

So, it looks like cluster logging operator should be fixed.
I would appreciate any comments or suggestions.

The detail is below.
I regenerated the certificates as follows.
```
[core@rhocp410-worker6 ~]$ sudo nsenter -t $(pgrep cluster-logging) --mount
[root@rhocp410-worker6 /]# cp -p /usr/bin/scripts/cert_generation.sh /usr/bin/scripts/cert_generation.sh.org^C
[root@rhocp410-worker6 /]# vi /usr/bin/scripts/cert_generation.sh
[root@rhocp410-worker6 /]# diff -u /usr/bin/scripts/cert_generation.sh.org /usr/bin/scripts/cert_generation.sh
--- /usr/bin/scripts/cert_generation.sh.org     2020-03-23 09:10:22.000000000 +0000
+++ /usr/bin/scripts/cert_generation.sh 2020-04-03 03:53:33.824057567 +0000
@@ -6,7 +6,7 @@
 NAMESPACE=$2
 CA_PATH=${CA_PATH:-$WORKING_DIR/ca.crt}
 LOG_STORE=$3
-REGENERATE_NEEDED=0
+REGENERATE_NEEDED=1

 function init_cert_files() {

@@ -20,7 +20,7 @@
 }

 function generate_signing_ca() {
-  if [ ! -f ${WORKING_DIR}/ca.crt ] || [ ! -f ${WORKING_DIR}/ca.key ] || ! openssl x509 -checkend 0 -noout -in ${WORKING_DIR}/ca.crt; then
+  if [ $REGENERATE_NEEDED -ne 0 ] || [ ! -f ${WORKING_DIR}/ca.crt ] || [ ! -f ${WORKING_DIR}/ca.key ] || ! openssl x509 -checkend 0 -noout -in ${WORKING_DIR}/ca.crt; then
     openssl req -x509 \
                 -new \
                 -newkey rsa:4096 \

Before Regeneration:

  [root@rhocp410-worker6 /]# cksum /tmp/ocp-clo/*
  ...
  3323181290 1850 /tmp/ocp-clo/ca.crt
  3600238151 2535 /tmp/ocp-clo/ca.db
  3849082317 20 /tmp/ocp-clo/ca.db.attr
  3849082317 20 /tmp/ocp-clo/ca.db.attr.old
  1917137176 2469 /tmp/ocp-clo/ca.db.old
  681125202 3272 /tmp/ocp-clo/ca.key
  1561848553 3 /tmp/ocp-clo/ca.serial.txt
  1557087854 3 /tmp/ocp-clo/ca.serial.txt.old
  ...

After Regeneration:

  [root@rhocp410-worker6 /]# cksum /tmp/ocp-clo/*
  ...
  3449268057 1850 /tmp/ocp-clo/ca.crt
  549398629 14196 /tmp/ocp-clo/ca.db
  3849082317 20 /tmp/ocp-clo/ca.db.attr
  3849082317 20 /tmp/ocp-clo/ca.db.attr.old
  3321310373 14130 /tmp/ocp-clo/ca.db.old
  652093969 3272 /tmp/ocp-clo/ca.key
  1218432204 3 /tmp/ocp-clo/ca.serial.txt
  1301580633 3 /tmp/ocp-clo/ca.serial.txt.old
  ...
  [root@rhocp410-worker6 /]# openssl x509 -in /tmp/ocp-clo/system.logging.fluentd.crt -text | grep Not
              Not Before: Apr  3 04:29:52 2020 GMT
              Not After : Apr  3 04:29:52 2022 GMT
```
Only elasticsearch pod was recreated after that.
```
[Fri Apr 03 13:30:18 root@api ~]# oc get pods -o wide
NAME                                           READY   STATUS        RESTARTS   AGE     IP            NODE                                             NOMINATED NODE   READINESS GATES
cluster-logging-operator-785b4dd44b-tc227      1/1     Running       0          18h     10.130.2.7    rhocp410-worker6.rhocp410.cluster.sub.nec.test   <none>           <none>
curator-1585884600-sng8l                       0/1     Error         0          60m     10.130.0.20   rhocp410-worker0.rhocp410.cluster.sub.nec.test   <none>           <none>
elasticsearch-cdm-pxnmds9o-1-6455fc7db-ml699   2/2     Terminating   0          5m14s   10.130.0.24   rhocp410-worker0.rhocp410.cluster.sub.nec.test   <none>           <none>
fluentd-t5vmb                                  1/1     Running       0          78m     10.131.2.18   rhocp410-worker5.rhocp410.cluster.sub.nec.test   <none>           <none>
fluentd-xv4vg                                  1/1     Running       0          78m     10.130.2.11   rhocp410-worker6.rhocp410.cluster.sub.nec.test   <none>           <none>
kibana-6ccd94b746-r4jdz                        2/2     Running       0          78m     10.129.0.23   rhocp410-worker2.rhocp410.cluster.sub.nec.test   <none>           <none>

[Fri Apr 03 13:30:46 root@api ~]# oc get pods -o wide
NAME                                           READY   STATUS              RESTARTS   AGE   IP            NODE                                             NOMINATED NODE   READINESS GATES
cluster-logging-operator-785b4dd44b-tc227      1/1     Running             0          19h   10.130.2.7    rhocp410-worker6.rhocp410.cluster.sub.nec.test   <none>           <none>
curator-1585884600-sng8l                       0/1     Error               0          60m   10.130.0.20   rhocp410-worker0.rhocp410.cluster.sub.nec.test   <none>           <none>
elasticsearch-cdm-pxnmds9o-1-6455fc7db-tw264   0/2     ContainerCreating   0          3s    <none>        rhocp410-worker0.rhocp410.cluster.sub.nec.test   <none>           <none>
fluentd-t5vmb                                  1/1     Running             0          78m   10.131.2.18   rhocp410-worker5.rhocp410.cluster.sub.nec.test   <none>           <none>
fluentd-xv4vg                                  1/1     Running             0          78m   10.130.2.11   rhocp410-worker6.rhocp410.cluster.sub.nec.test   <none>           <none>
kibana-6ccd94b746-r4jdz                        2/2     Running             0          78m   10.129.0.23   rhocp410-worker2.rhocp410.cluster.sub.nec.test   <none>           <none>```
```
Elasticsearch and kibana reloaded certificates, but fluentd is still using old certificate.
```
bash-4.2$ curl https://elasticsearch.openshift-logging.svc:9200 -vvvs 2>&1 | grep -i date
*       start date: Apr 03 04:29:56 2020 GMT
*       expire date: Apr 03 04:29:56 2022 GMT

bash-4.2$ curl https://fluentd.openshift-logging.svc.cluster.local:24231 -vvvs 2>&1 | grep -i date
*       start date: Mar 31 11:01:54 2020 GMT
*       expire date: Mar 31 11:01:55 2022 GMT

bash-4.2$ curl https://kibana.openshift-logging.svc.cluster.local -vvvs 2>&1 | grep -i date
*       start date: Apr 03 04:29:54 2020 GMT
*       expire date: Apr 03 04:29:54 2022 GMT
```
Kibana route became unavailable since it has old certificate.
```
[Fri Apr 03 13:38:49 root@api ~]# curl https://kibana-openshift-logging.apps.rhocp410.cluster.sub.nec.test -ks | grep available
      <h1>Application is not available</h1>
[Fri Apr 03 13:38:56 root@api ~]# oc get route -n openshift-logging kibana
NAME     HOST/PORT                                                     PATH   SERVICES   PORT    TERMINATION          WILDCARD
kibana   kibana-openshift-logging.apps.rhocp410.cluster.sub.nec.test          kibana     <all>   reencrypt/Redirect   None
[Fri Apr 03 13:39:00 root@api ~]# oc get route -n openshift-logging kibana -o jsonpath='{.spec.tls.caCertificate}'| openssl x509 -text | grep Not
            Not Before: Feb 19 11:13:50 2020 GMT
            Not After : Feb 17 11:13:50 2025 GMT
```

Reply:
@k-keiichi since this is focused specifically on the documentation of certificates can you open a new BZ for each type of rotation bug you're seeing?  That way it can be assigned to the appropriate team for direct comment.  

It looks like you need to open one for monitoring and one for logging.  Please include steps to recreate, and the info you posted above so the engineering teams can respond.  Thanks!

Reply:
@pweil 
Thanks. I will create new BZs for the above comments.